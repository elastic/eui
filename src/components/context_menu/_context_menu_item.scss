.euiContextMenuItem {
  padding: $euiSizeM;
  width: 100%;
  text-align: left;
  color: $euiTextColor;

  &:hover, &:focus {
    text-decoration: underline;
  }

  &:focus {
    background-color: $euiFocusBackgroundColor;
  }

  &.euiContextMenuItem-isDisabled {
    color: $euiButtonColorDisabled;
    cursor: default;

    &:hover, &:focus {
      text-decoration: none;
    }
  }
}

  .euiContextMenuItem__inner {
    display: flex;
  }

  .euiContextMenuItem__text {
    flex-grow: 1;
  }

  .euiContextMenuItem__arrow {
    align-self: flex-end;
  }
