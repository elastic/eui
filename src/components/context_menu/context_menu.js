import React, {
  Component,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { EuiContextMenuPanel } from './context_menu_panel';
import { EuiContextMenuItem } from './context_menu_item';

function mapIdsToPanels(panels) {
  const map = {};

  panels.forEach(panel => {
    map[panel.id] = panel;
  });

  return map;
}

function mapIdsToPreviousPanels(panels) {
  const idToPreviousPanelIdMap = {};

  panels.forEach(panel => {
    if (Array.isArray(panel.items)) {
      panel.items.forEach(item => {
        const isCloseable = item.panel !== undefined;
        if (isCloseable) {
          idToPreviousPanelIdMap[item.panel] = panel.id;
        }
      });
    }
  });

  return idToPreviousPanelIdMap;
}

function mapPanelItemsToPanels(panels) {
  const idAndItemIndexToPanelIdMap = {};

  panels.forEach(panel => {
    idAndItemIndexToPanelIdMap[panel.id] = {};

    if (panel.items) {
      panel.items.forEach((item, index) => {
        if (item.panel) {
          idAndItemIndexToPanelIdMap[panel.id][index] = item.panel;
        }
      });
    }
  });

  return idAndItemIndexToPanelIdMap;
}

export const EuiContextMenuPanelItemShape = PropTypes.shape({
  name: PropTypes.string,
  icon: PropTypes.node,
  onClick: PropTypes.func,
  // If given, shows the panel with this id when clicked:
  panel: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  disabled: PropTypes.bool,
});

export const EuiContextMenuPanelShape = PropTypes.shape({
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  width: PropTypes.number,  // Pixel value to set the panel width to
  content: PropTypes.node,  // Either content or items array should be given.
  items: PropTypes.arrayOf(EuiContextMenuPanelItemShape),
  title: PropTypes.string,
});

export class EuiContextMenu extends Component {
  static propTypes = {
    className: PropTypes.string,
    panels: PropTypes.arrayOf(EuiContextMenuPanelShape),
    initialPanelId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  };

  static defaultProps = {
    panels: [],
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    const { panels } = nextProps;

    if (prevState.prevProps.panels !== panels) {
      return {
        prevProps: { panels },
        idToPanelMap: mapIdsToPanels(panels),
        idToPreviousPanelIdMap: mapIdsToPreviousPanels(panels),
        idAndItemIndexToPanelIdMap: mapPanelItemsToPanels(panels)
      };
    }

    return null;
  }

  constructor(props) {
    super(props);

    this.state = {
      prevProps: {},
      idToPanelMap: {},
      idToPreviousPanelIdMap: {},
      idAndItemIndexToPanelIdMap: {},
      idToRenderedItemsMap: this.mapIdsToRenderedItems(this.props.panels),

      height: undefined,
      outgoingPanelId: undefined,
      incomingPanelId: props.initialPanelId,
      transitionDirection: undefined,
      isOutgoingPanelVisible: false,
      focusedItemIndex: undefined,
      isUsingKeyboardToNavigate: false,
    };
  }

  componentDidUpdate(prevProps) {
    if (prevProps.panels !== this.props.panels) {
      this.setState({ // eslint-disable-line react/no-did-update-set-state
        idToRenderedItemsMap: this.mapIdsToRenderedItems(this.props.panels),
      });
    }
  }

  hasPreviousPanel = panelId => {
    const previousPanelId = this.state.idToPreviousPanelIdMap[panelId];
    return typeof previousPanelId !== 'undefined';
  };

  showPanel(panelId, direction) {
    this.setState({
      outgoingPanelId: this.state.incomingPanelId,
      incomingPanelId: panelId,
      transitionDirection: direction,
      isOutgoingPanelVisible: true,
    });
  }

  showNextPanel = itemIndex => {
    const nextPanelId = this.state.idAndItemIndexToPanelIdMap[this.state.incomingPanelId][itemIndex];
    if (nextPanelId) {
      if (this.state.isUsingKeyboardToNavigate) {
        this.setState({
          focusedItemIndex: 0,
        });
      }

      this.showPanel(nextPanelId, 'next');
    }
  };

  showPreviousPanel = () => {
    // If there's a previous panel, then we can close the current panel to go back to it.
    if (this.hasPreviousPanel(this.state.incomingPanelId)) {
      const previousPanelId = this.state.idToPreviousPanelIdMap[this.state.incomingPanelId];

      // Set focus on the item which shows the panel we're leaving.
      const previousPanel = this.state.idToPanelMap[previousPanelId];
      const focusedItemIndex = previousPanel.items.findIndex(
        item => item.panel === this.state.incomingPanelId
      );

      if (focusedItemIndex !== -1) {
        this.setState({
          focusedItemIndex,
        });
      }

      this.showPanel(previousPanelId, 'previous');
    }
  };

  onIncomingPanelHeightChange = height => {
    this.setState(({ height: prevHeight }) => {
      if (height === prevHeight) {
        return null;
      } else {
        return { height };
      }
    });
  };

  onOutGoingPanelTransitionComplete = () => {
    this.setState({
      isOutgoingPanelVisible: false,
    });
  };

  onUseKeyboardToNavigate = () => {
    if (!this.state.isUsingKeyboardToNavigate) {
      this.setState({
        isUsingKeyboardToNavigate: true,
      });
    }
  };

  mapIdsToRenderedItems = panels => {
    const idToRenderedItemsMap = {};

    // Pre-rendering the items lets us check reference equality inside of EuiContextMenuPanel.
    panels.forEach(panel => {
      idToRenderedItemsMap[panel.id] = this.renderItems(panel.items);
    });

    return idToRenderedItemsMap;
  };

  renderItems(items = []) {
    return items.map((item, index) => {
      const {
        panel,
        name,
        icon,
        onClick,
        toolTipTitle,
        toolTipContent,
        ...rest
      } = item;

      const onClickHandler = panel
        ? (event) => {
          if (onClick && event) {
            event.persist();
          }
          // This component is commonly wrapped in a EuiOutsideClickDetector, which means we'll
          // need to wait for that logic to complete before re-rendering the DOM via showPanel.
          window.requestAnimationFrame(() => {
            if (onClick) onClick(event);
            this.showNextPanel(index);
          });
        } : onClick;

      return (
        <EuiContextMenuItem
          key={name}
          icon={icon}
          onClick={onClickHandler}
          hasPanel={Boolean(panel)}
          toolTipTitle={toolTipTitle}
          toolTipContent={toolTipContent}
          {...rest}
        >
          {name}
        </EuiContextMenuItem>
      );
    });
  }

  renderPanel(panelId, transitionType) {
    const panel = this.state.idToPanelMap[panelId];

    if (!panel) {
      return;
    }

    // As above, we need to wait for EuiOutsideClickDetector to complete its logic before
    // re-rendering via showPanel.
    let onClose;
    if (this.hasPreviousPanel(panelId)) {
      onClose = () => window.requestAnimationFrame(this.showPreviousPanel);
    }

    return (
      <EuiContextMenuPanel
        key={panelId}
        className="euiContextMenu__panel"
        onHeightChange={(transitionType === 'in') ? this.onIncomingPanelHeightChange : undefined}
        onTransitionComplete={(transitionType === 'out') ? this.onOutGoingPanelTransitionComplete : undefined}
        title={panel.title}
        onClose={onClose}
        transitionType={this.state.isOutgoingPanelVisible ? transitionType : undefined}
        transitionDirection={this.state.isOutgoingPanelVisible ? this.state.transitionDirection : undefined}
        hasFocus={transitionType === 'in'}
        items={this.state.idToRenderedItemsMap[panelId]}
        initialFocusedItemIndex={this.state.isUsingKeyboardToNavigate ? this.state.focusedItemIndex : undefined}
        onUseKeyboardToNavigate={this.onUseKeyboardToNavigate}
        showNextPanel={this.showNextPanel}
        showPreviousPanel={this.showPreviousPanel}
      >
        {panel.content}
      </EuiContextMenuPanel>
    );
  }

  render() {
    const {
      panels, // eslint-disable-line no-unused-vars
      className,
      initialPanelId, // eslint-disable-line no-unused-vars
      ...rest
    } = this.props;

    const incomingPanel = this.renderPanel(this.state.incomingPanelId, 'in');
    let outgoingPanel;

    if (this.state.isOutgoingPanelVisible) {
      outgoingPanel = this.renderPanel(this.state.outgoingPanelId, 'out');
    }

    const width =
      this.state.idToPanelMap[this.state.incomingPanelId] &&
      this.state.idToPanelMap[this.state.incomingPanelId].width ?
        this.state.idToPanelMap[this.state.incomingPanelId].width : undefined;

    const classes = classNames('euiContextMenu', className);

    return (
      <div
        className={classes}
        style={{ height: this.state.height, width: width }}
        {...rest}
      >
        {outgoingPanel}
        {incomingPanel}
      </div>
    );
  }
}
