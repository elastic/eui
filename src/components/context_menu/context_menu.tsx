/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  Component,
  HTMLAttributes,
  ReactElement,
  ReactNode,
} from 'react';
import classNames from 'classnames';

import { CommonProps, ExclusiveUnion, keysOf } from '../common';
import {
  EuiContextMenuPanel,
  EuiContextMenuPanelTransitionDirection,
  EuiContextMenuPanelTransitionType,
} from './context_menu_panel';
import {
  EuiContextMenuItem,
  EuiContextMenuItemProps,
} from './context_menu_item';
import { EuiHorizontalRule, EuiHorizontalRuleProps } from '../horizontal_rule';

export type EuiContextMenuPanelId = string | number;

export type EuiContextMenuPanelItemDescriptorEntry = Omit<
  EuiContextMenuItemProps,
  'hasPanel'
> & {
  name: React.ReactNode;
  key?: string;
  panel?: EuiContextMenuPanelId;
};

export interface EuiContextMenuPanelItemSeparator
  extends EuiHorizontalRuleProps {
  isSeparator: true;
  key?: string;
}

export type EuiContextMenuPanelItemDescriptor = ExclusiveUnion<
  EuiContextMenuPanelItemDescriptorEntry,
  EuiContextMenuPanelItemSeparator
>;

export interface EuiContextMenuPanelDescriptor {
  id: EuiContextMenuPanelId;
  title?: ReactNode;
  items?: EuiContextMenuPanelItemDescriptor[];
  content?: ReactNode;
  width?: number;
  initialFocusedItemIndex?: number;
  /**
   * Alters the size of the items and the title
   */
  size?: keyof typeof sizeToClassNameMap;
}

const sizeToClassNameMap = {
  s: 'euiContextMenu--small',
  m: null,
};

export const SIZES = keysOf(sizeToClassNameMap);

export type EuiContextMenuProps = CommonProps &
  Omit<HTMLAttributes<HTMLDivElement>, 'style'> & {
    panels?: EuiContextMenuPanelDescriptor[];
    initialPanelId?: EuiContextMenuPanelId;
    /**
     * Alters the size of the items and the title
     */
    size?: keyof typeof sizeToClassNameMap;
  };

const isItemSeparator = (
  item: EuiContextMenuPanelItemDescriptor
): item is EuiContextMenuPanelItemSeparator =>
  (item as EuiContextMenuPanelItemSeparator).isSeparator === true;

function mapIdsToPanels(panels: EuiContextMenuPanelDescriptor[]) {
  const map: { [id: string]: EuiContextMenuPanelDescriptor } = {};

  panels.forEach((panel) => {
    map[panel.id] = panel;
  });

  return map;
}

function mapIdsToPreviousPanels(panels: EuiContextMenuPanelDescriptor[]) {
  const idToPreviousPanelIdMap: { [panel: string]: EuiContextMenuPanelId } = {};

  panels.forEach((panel) => {
    if (Array.isArray(panel.items)) {
      panel.items.forEach((item) => {
        if (isItemSeparator(item)) return;
        const isCloseable = item.panel !== undefined;
        if (isCloseable) {
          idToPreviousPanelIdMap[item.panel!] = panel.id;
        }
      });
    }
  });

  return idToPreviousPanelIdMap;
}

function mapPanelItemsToPanels(panels: EuiContextMenuPanelDescriptor[]) {
  const idAndItemIndexToPanelIdMap: {
    [id: string]: { [index: string]: EuiContextMenuPanelId };
  } = {};

  panels.forEach((panel) => {
    idAndItemIndexToPanelIdMap[panel.id] = {};

    if (panel.items) {
      panel.items.forEach((item, index) => {
        if (isItemSeparator(item)) return;
        if (item.panel) {
          idAndItemIndexToPanelIdMap[panel.id][index] = item.panel;
        }
      });
    }
  });

  return idAndItemIndexToPanelIdMap;
}

interface State {
  prevProps: {
    panels?: EuiContextMenuPanelDescriptor[];
  };
  idToPanelMap: { [id: string]: EuiContextMenuPanelDescriptor };
  idToPreviousPanelIdMap: { [panel: string]: EuiContextMenuPanelId };
  idAndItemIndexToPanelIdMap: {
    [id: string]: { [index: string]: EuiContextMenuPanelId };
  };
  idToRenderedItemsMap: { [id: string]: ReactElement[] };

  height?: number;
  outgoingPanelId?: EuiContextMenuPanelId;
  incomingPanelId?: EuiContextMenuPanelId;
  transitionDirection?: EuiContextMenuPanelTransitionDirection;
  isOutgoingPanelVisible: boolean;
  focusedItemIndex?: number;
  isUsingKeyboardToNavigate: boolean;
}

export class EuiContextMenu extends Component<EuiContextMenuProps, State> {
  static defaultProps: Partial<EuiContextMenuProps> = {
    panels: [],
    size: 'm',
  };

  static getDerivedStateFromProps(
    nextProps: EuiContextMenuProps,
    prevState: State
  ): Partial<State> | null {
    const { panels } = nextProps;

    if (panels && prevState.prevProps.panels !== panels) {
      return {
        prevProps: { panels },
        idToPanelMap: mapIdsToPanels(panels),
        idToPreviousPanelIdMap: mapIdsToPreviousPanels(panels),
        idAndItemIndexToPanelIdMap: mapPanelItemsToPanels(panels),
      };
    }

    return null;
  }

  constructor(props: EuiContextMenuProps) {
    super(props);

    this.state = {
      prevProps: {},
      idToPanelMap: {},
      idToPreviousPanelIdMap: {},
      idAndItemIndexToPanelIdMap: {},
      idToRenderedItemsMap: this.mapIdsToRenderedItems(this.props.panels),

      height: undefined,
      outgoingPanelId: undefined,
      incomingPanelId: props.initialPanelId,
      transitionDirection: undefined,
      isOutgoingPanelVisible: false,
      focusedItemIndex: undefined,
      isUsingKeyboardToNavigate: false,
    };
  }

  componentDidUpdate(prevProps: EuiContextMenuProps) {
    if (prevProps.panels !== this.props.panels) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        idToRenderedItemsMap: this.mapIdsToRenderedItems(this.props.panels),
      });
    }
  }

  hasPreviousPanel = (panelId: EuiContextMenuPanelId) => {
    const previousPanelId = this.state.idToPreviousPanelIdMap[panelId];
    return typeof previousPanelId !== 'undefined';
  };

  showPanel(
    panelId: EuiContextMenuPanelId,
    direction?: EuiContextMenuPanelTransitionDirection
  ) {
    this.setState({
      outgoingPanelId: this.state.incomingPanelId,
      incomingPanelId: panelId,
      transitionDirection: direction,
      isOutgoingPanelVisible: true,
    });
  }

  showNextPanel = (itemIndex?: number) => {
    if (itemIndex == null) {
      return;
    }

    const nextPanelId = this.state.idAndItemIndexToPanelIdMap[
      this.state.incomingPanelId!
    ][itemIndex];

    if (nextPanelId) {
      if (this.state.isUsingKeyboardToNavigate) {
        this.setState(({ idToPanelMap }) => ({
          focusedItemIndex:
            idToPanelMap[nextPanelId].initialFocusedItemIndex ?? 0,
        }));
      }

      this.showPanel(nextPanelId, 'next');
    }
  };

  showPreviousPanel = () => {
    // If there's a previous panel, then we can close the current panel to go back to it.
    if (this.hasPreviousPanel(this.state.incomingPanelId!)) {
      const previousPanelId = this.state.idToPreviousPanelIdMap[
        this.state.incomingPanelId!
      ];

      // Set focus on the item which shows the panel we're leaving.
      const previousPanel = this.state.idToPanelMap[previousPanelId];
      const focusedItemIndex = previousPanel.items!.findIndex(
        (item) =>
          !isItemSeparator(item) && item.panel === this.state.incomingPanelId
      );

      if (focusedItemIndex !== -1) {
        this.setState({
          focusedItemIndex,
        });
      }

      this.showPanel(previousPanelId, 'previous');
    }
  };

  onIncomingPanelHeightChange = (height: number) => {
    this.setState(({ height: prevHeight }) => {
      if (height === prevHeight) {
        return null;
      }

      return { height };
    });
  };

  onOutGoingPanelTransitionComplete = () => {
    this.setState({
      isOutgoingPanelVisible: false,
    });
  };

  onUseKeyboardToNavigate = () => {
    if (!this.state.isUsingKeyboardToNavigate) {
      this.setState({
        isUsingKeyboardToNavigate: true,
      });
    }
  };

  mapIdsToRenderedItems = (panels: EuiContextMenuPanelDescriptor[] = []) => {
    const idToRenderedItemsMap: { [id: string]: ReactElement[] } = {};

    // Pre-rendering the items lets us check reference equality inside of EuiContextMenuPanel.
    panels.forEach((panel) => {
      idToRenderedItemsMap[panel.id] = this.renderItems(panel.items);
    });

    return idToRenderedItemsMap;
  };

  renderItems(items: EuiContextMenuPanelItemDescriptor[] = []) {
    return items.map((item, index) => {
      if (isItemSeparator(item)) {
        const { isSeparator: omit, key = index, ...rest } = item;
        return <EuiHorizontalRule key={key} margin="none" {...rest} />;
      }

      const {
        panel,
        name,
        key,
        icon,
        onClick,
        toolTipTitle,
        toolTipContent,
        ...rest
      } = item;

      const onClickHandler = panel
        ? (event: React.MouseEvent) => {
            if (onClick && event) {
              event.persist();
            }
            // This component is commonly wrapped in a EuiOutsideClickDetector, which means we'll
            // need to wait for that logic to complete before re-rendering the DOM via showPanel.
            window.requestAnimationFrame(() => {
              if (onClick) {
                onClick(event);
              }
              this.showNextPanel(index);
            });
          }
        : onClick;

      return (
        <EuiContextMenuItem
          key={key || (typeof name === 'string' ? name : undefined) || index}
          icon={icon}
          onClick={onClickHandler}
          hasPanel={Boolean(panel)}
          toolTipTitle={toolTipTitle}
          toolTipContent={toolTipContent}
          {...rest}>
          {name}
        </EuiContextMenuItem>
      );
    });
  }

  renderPanel(
    panelId: EuiContextMenuPanelId,
    transitionType: EuiContextMenuPanelTransitionType
  ) {
    const panel = this.state.idToPanelMap[panelId];

    if (!panel) {
      return;
    }

    // As above, we need to wait for EuiOutsideClickDetector to complete its logic before
    // re-rendering via showPanel.
    let onClose;
    if (this.hasPreviousPanel(panelId)) {
      onClose = () => window.requestAnimationFrame(this.showPreviousPanel);
    }

    return (
      <EuiContextMenuPanel
        key={panelId}
        size={this.props.size}
        className="euiContextMenu__panel"
        onHeightChange={
          transitionType === 'in' ? this.onIncomingPanelHeightChange : undefined
        }
        onTransitionComplete={
          transitionType === 'out'
            ? this.onOutGoingPanelTransitionComplete
            : undefined
        }
        title={panel.title}
        onClose={onClose}
        transitionType={
          this.state.isOutgoingPanelVisible ? transitionType : undefined
        }
        transitionDirection={
          this.state.isOutgoingPanelVisible
            ? this.state.transitionDirection
            : undefined
        }
        hasFocus={transitionType === 'in'}
        items={this.state.idToRenderedItemsMap[panelId]}
        initialFocusedItemIndex={
          this.state.isUsingKeyboardToNavigate
            ? this.state.focusedItemIndex
            : panel.initialFocusedItemIndex
        }
        onUseKeyboardToNavigate={this.onUseKeyboardToNavigate}
        showNextPanel={this.showNextPanel}
        showPreviousPanel={this.showPreviousPanel}>
        {panel.content}
      </EuiContextMenuPanel>
    );
  }

  render() {
    const { panels, className, initialPanelId, size, ...rest } = this.props;

    const incomingPanel = this.renderPanel(this.state.incomingPanelId!, 'in');
    let outgoingPanel;

    if (this.state.isOutgoingPanelVisible) {
      outgoingPanel = this.renderPanel(this.state.outgoingPanelId!, 'out');
    }

    const width =
      this.state.idToPanelMap[this.state.incomingPanelId!] &&
      this.state.idToPanelMap[this.state.incomingPanelId!].width
        ? this.state.idToPanelMap[this.state.incomingPanelId!].width
        : undefined;

    const classes = classNames(
      'euiContextMenu',
      size && sizeToClassNameMap[size],
      className
    );

    return (
      <div
        className={classes}
        style={{ height: this.state.height, width: width }}
        {...rest}>
        {outgoingPanel}
        {incomingPanel}
      </div>
    );
  }
}
