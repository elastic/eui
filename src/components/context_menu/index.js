export {
  EuiContextMenu,
} from './context_menu';

export {
  EuiContextMenuPanel,
} from './context_menu_panel';

export {
  EuiContextMenuItem,
} from './context_menu_item';
