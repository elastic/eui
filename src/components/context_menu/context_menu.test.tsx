/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { render, mount } from 'enzyme';
import { requiredProps, takeMountedSnapshot } from '../../test';

import { EuiContextMenu, SIZES } from './context_menu';
import { setTimeout } from 'timers';

const panel3 = {
  id: 3,
  title: '3',
  items: [
    {
      name: <span style={{ color: 'tomato' }}>foo</span>,
      key: 'foo',
    },
  ],
};

const panel2 = {
  id: 2,
  title: '2',
  content: <div>2</div>,
};

const panel1 = {
  id: 1,
  title: '1',
  width: 400,
  items: [
    {
      name: '2a',
      panel: 2,
    },
    {
      name: '2b',
      panel: 2,
    },
    {
      name: '2c',
      panel: 2,
    },
  ],
};

const panel0 = {
  id: 0,
  title: '0',
  items: [
    {
      name: '1',
      panel: 1,
    },
  ],
};

const panels = [panel0, panel1, panel2, panel3];

export const tick = (ms = 0) =>
  new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

describe('EuiContextMenu', () => {
  test('is rendered', () => {
    const component = render(<EuiContextMenu {...requiredProps} />);

    expect(component).toMatchSnapshot();
  });

  it('panel item can contain JSX', () => {
    const component = render(
      <EuiContextMenu panels={panels} initialPanelId={3} />
    );

    expect(component).toMatchSnapshot();
  });

  it('panel item can be a separator line', () => {
    const component = render(
      <EuiContextMenu
        panels={[
          {
            id: 3,
            title: 'Testing separator',
            items: [
              { name: 'Foo', key: 'foo' },
              { isSeparator: true },
              { name: 'Bar', key: 'bar' },
            ],
          },
        ]}
        initialPanelId={3}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('can pass-through horizontal rule props', () => {
    const component = render(
      <EuiContextMenu
        panels={[
          {
            id: 3,
            title: 'Testing separator',
            items: [
              {
                isSeparator: true,
                key: 'separator',
                margin: 's',
                size: 'half',
              },
            ],
          },
        ]}
        initialPanelId={3}
      />
    );

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    describe('panels and initialPanelId', () => {
      it('renders the referenced panel', () => {
        const component = render(
          <EuiContextMenu panels={panels} initialPanelId={2} />
        );

        expect(component).toMatchSnapshot();
      });

      it('allows you to click the title button to go back to the previous panel', async () => {
        const component = mount(
          <EuiContextMenu panels={panels} initialPanelId={2} />
        );

        await tick(20);

        expect(takeMountedSnapshot(component)).toMatchSnapshot();

        // Navigate to a different panel.
        component
          .find('[data-test-subj="contextMenuPanelTitleButton"]')
          .simulate('click');

        await tick(20);

        expect(takeMountedSnapshot(component)).toMatchSnapshot();
      });
    });

    describe('size', () => {
      SIZES.forEach((size) => {
        it(`${size} is rendered`, () => {
          const component = render(
            <EuiContextMenu panels={panels} initialPanelId={2} size={size} />
          );

          expect(component).toMatchSnapshot();
        });
      });
    });
  });
});
