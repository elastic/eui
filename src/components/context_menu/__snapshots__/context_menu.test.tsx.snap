// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiContextMenu can pass-through horizontal rule props 1`] = `
<div
  class="euiContextMenu"
>
  <div
    class="euiContextMenuPanel euiContextMenu__panel"
    tabindex="-1"
  >
    <div
      class="euiContextMenuPanelTitle"
    >
      <span
        class="euiContextMenu__itemLayout"
      >
        Testing separator
      </span>
    </div>
    <div>
      <div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--half euiHorizontalRule--marginSmall"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiContextMenu is rendered 1`] = `
<div
  aria-label="aria-label"
  class="euiContextMenu testClass1 testClass2"
  data-test-subj="test subject string"
/>
`;

exports[`EuiContextMenu panel item can be a separator line 1`] = `
<div
  class="euiContextMenu"
>
  <div
    class="euiContextMenuPanel euiContextMenu__panel"
    tabindex="-1"
  >
    <div
      class="euiContextMenuPanelTitle"
    >
      <span
        class="euiContextMenu__itemLayout"
      >
        Testing separator
      </span>
    </div>
    <div>
      <div>
        <button
          class="euiContextMenuItem"
          type="button"
        >
          <span
            class="euiContextMenu__itemLayout"
          >
            <span
              class="euiContextMenuItem__text"
            >
              Foo
            </span>
          </span>
        </button>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full"
        />
        <button
          class="euiContextMenuItem"
          type="button"
        >
          <span
            class="euiContextMenu__itemLayout"
          >
            <span
              class="euiContextMenuItem__text"
            >
              Bar
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiContextMenu panel item can contain JSX 1`] = `
<div
  class="euiContextMenu"
>
  <div
    class="euiContextMenuPanel euiContextMenu__panel"
    tabindex="-1"
  >
    <div
      class="euiContextMenuPanelTitle"
    >
      <span
        class="euiContextMenu__itemLayout"
      >
        3
      </span>
    </div>
    <div>
      <div>
        <button
          class="euiContextMenuItem"
          type="button"
        >
          <span
            class="euiContextMenu__itemLayout"
          >
            <span
              class="euiContextMenuItem__text"
            >
              <span
                style="color:tomato"
              >
                foo
              </span>
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiContextMenu props panels and initialPanelId allows you to click the title button to go back to the previous panel 1`] = `
<div
  class="euiContextMenu"
  style="height: 0px;"
>
  <div
    class="euiContextMenuPanel euiContextMenu__panel"
    tabindex="-1"
  >
    <button
      class="euiContextMenuPanelTitle"
      data-test-subj="contextMenuPanelTitleButton"
      type="button"
    >
      <span
        class="euiContextMenu__itemLayout"
      >
        <span
          class="euiContextMenu__icon"
          data-euiicon-type="arrowLeft"
        />
        <span
          class="euiContextMenu__text"
        >
          2
        </span>
      </span>
    </button>
    <div>
      <div>
        <div>
          2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiContextMenu props panels and initialPanelId allows you to click the title button to go back to the previous panel 2`] = `
<div
  class="euiContextMenu"
  style="height: 0px; width: 400px;"
>
  <div
    class="euiContextMenuPanel euiContextMenu__panel euiContextMenuPanel-txOutRight"
    tabindex="-1"
  >
    <button
      class="euiContextMenuPanelTitle"
      data-test-subj="contextMenuPanelTitleButton"
      type="button"
    >
      <span
        class="euiContextMenu__itemLayout"
      >
        <span
          class="euiContextMenu__icon"
          data-euiicon-type="arrowLeft"
        />
        <span
          class="euiContextMenu__text"
        >
          2
        </span>
      </span>
    </button>
    <div>
      <div>
        <div>
          2
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiContextMenuPanel euiContextMenu__panel euiContextMenuPanel-txInRight"
    tabindex="-1"
  >
    <button
      class="euiContextMenuPanelTitle"
      data-test-subj="contextMenuPanelTitleButton"
      type="button"
    >
      <span
        class="euiContextMenu__itemLayout"
      >
        <span
          class="euiContextMenu__icon"
          data-euiicon-type="arrowLeft"
        />
        <span
          class="euiContextMenu__text"
        >
          1
        </span>
      </span>
    </button>
    <div>
      <div>
        <button
          class="euiContextMenuItem"
          type="button"
        >
          <span
            class="euiContextMenu__itemLayout"
          >
            <span
              class="euiContextMenuItem__text"
            >
              2a
            </span>
            <span
              class="euiContextMenu__arrow"
              data-euiicon-type="arrowRight"
            />
          </span>
        </button>
        <button
          class="euiContextMenuItem"
          type="button"
        >
          <span
            class="euiContextMenu__itemLayout"
          >
            <span
              class="euiContextMenuItem__text"
            >
              2b
            </span>
            <span
              class="euiContextMenu__arrow"
              data-euiicon-type="arrowRight"
            />
          </span>
        </button>
        <button
          class="euiContextMenuItem"
          type="button"
        >
          <span
            class="euiContextMenu__itemLayout"
          >
            <span
              class="euiContextMenuItem__text"
            >
              2c
            </span>
            <span
              class="euiContextMenu__arrow"
              data-euiicon-type="arrowRight"
            />
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiContextMenu props panels and initialPanelId renders the referenced panel 1`] = `
<div
  class="euiContextMenu"
>
  <div
    class="euiContextMenuPanel euiContextMenu__panel"
    tabindex="-1"
  >
    <button
      class="euiContextMenuPanelTitle"
      data-test-subj="contextMenuPanelTitleButton"
      type="button"
    >
      <span
        class="euiContextMenu__itemLayout"
      >
        <span
          class="euiContextMenu__icon"
          data-euiicon-type="arrowLeft"
        />
        <span
          class="euiContextMenu__text"
        >
          2
        </span>
      </span>
    </button>
    <div>
      <div>
        <div>
          2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiContextMenu props size m is rendered 1`] = `
<div
  class="euiContextMenu"
>
  <div
    class="euiContextMenuPanel euiContextMenu__panel"
    tabindex="-1"
  >
    <button
      class="euiContextMenuPanelTitle"
      data-test-subj="contextMenuPanelTitleButton"
      type="button"
    >
      <span
        class="euiContextMenu__itemLayout"
      >
        <span
          class="euiContextMenu__icon"
          data-euiicon-type="arrowLeft"
        />
        <span
          class="euiContextMenu__text"
        >
          2
        </span>
      </span>
    </button>
    <div>
      <div>
        <div>
          2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EuiContextMenu props size s is rendered 1`] = `
<div
  class="euiContextMenu euiContextMenu--small"
>
  <div
    class="euiContextMenuPanel euiContextMenu__panel"
    tabindex="-1"
  >
    <button
      class="euiContextMenuPanelTitle euiContextMenuPanelTitle--small"
      data-test-subj="contextMenuPanelTitleButton"
      type="button"
    >
      <span
        class="euiContextMenu__itemLayout"
      >
        <span
          class="euiContextMenu__icon"
          data-euiicon-type="arrowLeft"
        />
        <span
          class="euiContextMenu__text"
        >
          2
        </span>
      </span>
    </button>
    <div>
      <div>
        <div>
          2
        </div>
      </div>
    </div>
  </div>
</div>
`;
